### ISSUE TITLE
Wait for a command to stop before shutting down

### ISSUE THREAD

**Issue by schomatis** [ipfs/go-ipfs#5002]
This problem was manifested in #4820 which I'll briefly describe here for context but actually exceeds its use case: while using the Badger datastore in offline mode, if the user cancels an `add` command there is a (very likely) possibility that the Badger DB is closed *while* open transactions are still operating on it (causing a panic as the Badger API does not allow such scenario but also doesn't take measures to prevent it, leaving that responsibility to the consumer of the API). The user is pressing `Ctrl-C` triggering the `IpfsNode.teardown()` procedure which proceeds to close the `BlockService` (and its inherent `Datastore` which is backed in this case by a Badger DB). In the meanwhile the `add` command keeps issuing store requests (which conflict with the close request from the `BlockService` shutdown causing the panic). Ideally the shutdown procedure should wait on the `add` command to finish (which also received a cancellation signal from the `Ctrl-C` but might take more time to act on it than `teardown()`) before closing the services on which the command relies.

As @Kubuxu mentioned this problem touches on other already existing issues (and I think it can have potential repercussions on others as well) so while my primary focus is to provide a solution for #4820 I'm opened to a more a comprehensive solution to improve inter-service communications (I'm considering the command dispatching a service although I'm not sure if it has that status).

I'm keeping this issue as the central reference for other issues/PRs addressing it (some of the will be in the `go-ipfs-cmds` repo).

**Comment by schomatis** [comment]
@Kubuxu Exactly, the `ipfs shutdown` command in online mode is the analog of `Ctrl-C`ing during command execution in offline mode, and the lack of communication between the core node shutdown and the service dispatching the HTTP response for the CLI client triggers a similar error (cause) to the Badger scenario.

**Comment by schomatis** [comment]
As a first step towards a solution, as [mentioned](https://github.com/ipfs/go-ipfs/issues/4820#issuecomment-376687766) in #4820, the [ReqLog](https://github.com/ipfs/go-ipfs/blob/master/commands/reqlog.go#L30) can be extended to cover the offline case (see https://github.com/ipfs/go-ipfs-cmds/issues/109) and have it as a bookkeeping control structure to know when a command has finished. (I'm not sure if we are abusing this structure way beyond it's original informative purpose without the responsibility of managing control flow, was that the original intention for the request log?)

**Comment by lidel** [comment]
Closing this. The original trigger (#4820) was badger v1 panicking when Ctrl-C during ipfs add closed the DB while transactions were still in flight. Badger v1 is now deprecated (#11187) and being removed (#11186), so that panic scenario goes away entirely.

  Since 2018, the shutdown path has improved:
  - Node teardown uses fx lifecycle hooks that close components in reverse dependency order
  - The HTTP server calls server.Shutdown() with a 30-second timeout, waiting for active handlers to finish
  - Commands propagate cancellation via contexts, so in-flight work winds down before services close
  - flatfs (the default) is resilient to this class of race -- each block is an independent file written via atomic os.Rename(), no shared DB handle to invalidate

  The general "wait for commands before teardown" concern was never a problem in practice with flatfs/leveldb. If it surfaces with a future datastore, it should be filed fresh with current context.
